File: 20_11_11/514_freedom_road.py
class Solution(object):
    def findRotateSteps(self, ring, key):
        """
        :type ring: str
        :type key: str
        :rtype: int
        """

        def distance(a, b):
            return min(abs(a-b), abs(len(ring)-abs(a-b)))+1
        
        index_dict = {}
        for n, i in enumerate(ring):
            if i in index_dict:
                index_dict[i].append(n)
            else:
                index_dict[i] = [n]
        
        MAX = 1e9
        dp = {0: 0}
        for key_i in key:
            dp_next = {}
            for key_n in index_dict[key_i]:
                dp_next[key_n] = MAX
                for loc, step in dp.items():
                    dp_next[key_n] = min(dp_next[key_n], distance(loc, key_n)+step)

            dp = dp_next
                
        return min(dp.values())


if __name__ == '__main__':
    solution = Solution()
    print(solution.findRotateSteps("godding", "gd"))

File: 20_11_11/514_freedom_road.cpp
#include<cstdio>
#include<iostream>
#include<algorithm>
#include<vector>
#include<map>
using namespace std;

class Solution {
public:
    int findRotateSteps(string ring, string key) {
        const int MAX = 1e9;
        int dp[110][110];
        int Nr = ring.size();
        int Nk = key.size();
        
        for(int i=0; i<Nk; i++){
            for(int j=0; j<Nr; j++){
                dp[i][j] = MAX;
            }
        }
        
        return 10;
    }
    

};

int main(){
    Solution solution;
    cout << solution.findRotateSteps("ring", "key"); 
}

